Error: branches platform index mapped for cond with linear transpose

### Description

The following error occurs with JAX 0.6.1. No error is thrown with JAX 0.6.0. 

The MWE reproduces the error. The gradient-via-linear transpose seems to be necessary to trigger the error - it does not have to be used and it may be deleted, but I cannot remove it. Similarly, if the linearisation is moved into the branch itself, the error disappears. 

For the diagonal - this was reported to be necessary by the user who first raised the issue, I poked around a little bit and just indexing into `z` does not seem to be enough.

```python
import jax
import jax.numpy as jnp


def inner_fn(x):
    z = jnp.outer(x, jnp.ones((10,)))
    return jnp.sum(jnp.diagonal(z))

def step(fn, x):
    f_eval, lin_fn = jax.linearize(fn, x)

    def with_transpose(args):
        grad = jax.linear_transpose(lin_fn, jnp.zeros_like(x))(1.0)
        del grad
        return jnp.zeros_like(x)
    
    def without_transpose(args):
        return jnp.zeros_like(x)
    
    return jax.lax.cond(f_eval > 0.1, with_transpose, without_transpose, None)

def outer_fn(x):
    return step(inner_fn, x)

jax.vmap(outer_fn)(jnp.ones((5, 10)))

# JaxStackTraceBeforeTransformation: AssertionError: The index of a cond with
# branches_platforms should be a platform_index and should never be mapped
```

Downstream issue: https://github.com/patrick-kidger/optimistix/issues/142

### System info (python version, jaxlib version, accelerator, etc.)

jax:    0.6.1
jaxlib: 0.6.1
numpy:  2.3.0
python: 3.13.3 (main, Apr  8 2025, 13:54:08) [Clang 16.0.0 (clang-1600.0.26.6)]
device info: cpu-1, 1 local devices"
process_count: 1
platform: uname_result(system='Darwin', node='[REDACTED]', release='24.5.0', version='Darwin Kernel Version 24.5.0: Tue Apr 22 19:54:26 PDT 2025; root:xnu-11417.121.6~2/RELEASE_ARM64_T8112', machine='arm64')

--- a/jax/_src/interpreters/ad.py
+++ b/jax/_src/interpreters/ad.py
@@ -277,7 +277,7 @@ def linearize(traceable: lu.WrappedFun,
     raise ValueError(
         "Linearization failed to produce known values for all output primals. "
         "This is typically caused by attempting to differentiate a function "
-        "uses an operation that does not support reverse-mode autodiff.")
+        "using an operation that does not support reverse-mode autodiff.")
   out_primals_consts = [pval.get_known() for pval in out_primals_pvals]
   if not has_aux:
     return out_primals_consts, out_tangents_pvals, jaxpr, consts

--- a/jax/_src/lax/control_flow/conditionals.py
+++ b/jax/_src/lax/control_flow/conditionals.py
@@ -490,11 +490,13 @@ def _cond_batching_rule(axis_data, args, dims, *, branches, **params):
     raise NotImplementedError(
         "IO effect not supported in vmap-of-cond.")
 
+  if "branches_platforms" in params and (index_dim is not batching.not_mapped):
+    # If we end up with a mapped index for a platform_dependent cond, we can
+    # replace the index with a fresh call to platform_index. See #29329.
+    index = platform_index_p.bind(platforms=params["branches_platforms"])
+    index_dim = batching.not_mapped
 
   if index_dim is not batching.not_mapped:
-    assert "branches_platforms" not in params, (
-        "The index of a cond with branches_platforms should be a "
-        "platform_index and should never be mapped")
     # Convert to a lax.select. While we could get away with not broadcasting
     # some operands yet, because all outputs must be broadcast together anyway
     # for the select we broadcast the input operands for simplicity and leave
@@ -563,7 +565,7 @@ def _cond_jvp(primals, tangents, *, branches, **params):
   return out_primals, out_tangents
 
 def _cond_partial_eval(trace, *tracers, branches, **params):
-  in_unknowns = [t.pval[0] is not None for t in tracers]
+  in_unknowns = [not t.pval.is_known() for t in tracers]
   index_uk, *ops_uk = in_unknowns
   if any(isinstance(eff, RefEffect) for branch in branches for eff in
       branch.jaxpr.effects):

--- a/tests/lax_control_flow_test.py
+++ b/tests/lax_control_flow_test.py
@@ -3072,6 +3072,26 @@ class LaxControlFlowTest(jtu.JaxTestCase):
     self.assertEqual(expect_a_dot, " dot(" in hlo)
     self.assertEqual(not expect_a_dot, " while(" in hlo)
 
+  def test_issue_29329(self):
+
+    def outer_fn(x):
+      def inner_fn(x):
+        return jax.jit(
+            lambda x: lax.platform_dependent(x,
+                                             default=jnp.sin,
+                                             other=jnp.cos))(x)
+
+      _, lin_fn = jax.linearize(inner_fn, x)
+
+      def with_transpose(x):
+        grad = jax.linear_transpose(lin_fn, x)(x)
+        del grad
+        return x
+
+      return jax.lax.cond(x[0][0] > 0., with_transpose, lambda x: x, x)
+
+    jax.vmap(outer_fn)(jnp.ones((5, 10, 10)))
+
   def test_scan_lowering_doesnt_introduce_singleton(self):
     b = 4
     i = 2

